tests/test_help_pages.py

```diff
@@ -26,10 +27,30 @@ def getter(self):
 
 
 @pytest.mark.help_page
-@pytest.mark.parametrize("args", traverse_commands(main, []), ids=" ".join)
-def test_access_help(no_api, args):
+@pytest.mark.parametrize("command_args", traverse_commands(main, []), ids=lambda x: " ".join(x[1]))
+def test_access_help(no_api, command_args):
     """Test, that all help screens are accessible without touching the api property."""
     runner = CliRunner()
+
+    command, args = command_args
     result = runner.invoke(main, args + ["--help"], catch_exceptions=False)
+
+    if result.exit_code != 0:
+        # verify whether a non-default repository context is necessary for executing the command
+        for context_name, registered_context_class in registered_repository_contexts.items():
+            for context_class in command.allowed_with_contexts:
```

Comment:
This works well for repository commands, but doesn't help for other multi-context commands like content. Maybe we need a giant table to register all the different types possible for each command.

Reply:
Yes, as you are saying. I focused only on repository commands. I did not want to create the table. However, it seems like it would be the best option.

Reply:
Can we get away with monkeypatching (maybe the `allowed_with_context` machinery) somehow?
All we need is the to see the help pages. I don't think, we need a fully working context. And i'd prefer not to add functionality to the actual code for the purpose of this test.

Reply:
> I don not quite remember the details of the issue we are tying to solve. But it is that some subcommands do not work with certain contexts, right?

Yes. In particular, the `pulp container repository remove-image` command is implemented only for push repositories (i.e., we have to always use the option `-t push` to change the default context and make the help pages available).

> If while enumerating subcommands, we clear their allowed_with_contexts attribute (with monkeypatch) they should again be available to the test, right?

Well, yes and no. We want to tell the test that this specific command can be run only within this specific context. In the PR, we are iterating through allowed context types until we reach the point where the help page is eventually available under one of the contexts.

> The question to follow up on: Is this then still testing the thing we want to test? Namely that all help pages are accessible without calling into the API.

I would say *yes*. Polymorphic commands could have different types of options per context. And, we want to show that to a user. When reading `allowed_with_context` classes, do we initiate any API calls?

Reply:
From reading that attribute, no we do not call the api. I guess it's hard to impossible to derive the proper value for `--type` from the context class in `allowed_with_context`, no?
Also by that logic, we should exercise all combinations of `--type` options.

How about the following: We raise a more specific exception (a new subclass of `ClickException` maybe called `IncompatibleContext`) on this condition (a small to unnoticeable codechange), iterate over all `--type` choices and ignore that new exception.

Reply:
I cannot think of any easy way of getting the all `--type` choices apart from doing introspection. Is there any Click facility that will allow me to do that? I suppose we have the same thing in our minds (i.e., the `--type` choices available under a corresponding decorator):
```
@pulp_group()
@type_option(
    choices={
        "container": PulpContainerRepositoryContext,
        "push": PulpContainerPushRepositoryContext,
    },
    default="container",
)
def repository() -> None:
    pass
```